### ISSUE TITLE
Make sure it's possible to catch all exceptions

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#2049]
We want to be able to implement `bracket` and `finally`. There's exceptions in pure code, due to calling `bug` or `todo` or an incomplete pattern match, and then also async exceptions due to a thread being killed.

Needs some more thought about how to handle. I think it could be good if all these things could be caught by an `Exception` handler, where:

```
ability Exception where raise : Failure -> x
```

@dolio let's discuss this

**Comment by pchiusano** [comment]
@dolio so my proposal is this - `bug` and `todo` should behave like this:

```
unique type Bug =
unique type Todo =

bug : x -> y
bug x = 
  f x = Exception.raise (Failure (typeLink Bug)) "bug" (Any x)
  coerceAbilities f x

todo : x -> y
todo x = 
  f x = Exception.raise (Failure (typeLink Todo)) "todo" (Any x)
  coerceAbilities f x
```

Now these can be caught in `Exception`.

**Comment by dolio** [comment]
This would work with the current ability implementation. But if we ever switch to a system where ability usage corresponds to under-the-hood information passing, then these `coerceAbilities` don't actually make any sense. They'll just produce invalid code.

So, it seems like this walks us more toward not fixing the ability soundness issues.

**Comment by pchiusano** [comment]
What we could do in that situation is just have `Exception.catch` be defined to also catch the exception thrown by `bug` and `todo`. 

But `Exception.catch` would be compiled to a builtin that gets selected when compiling a definition with the same hash as the current [`Exception.catch`](https://share.unison-lang.org/latest/namespaces/unison/base/;/terms/Exception/catch). This wouldn't box us in re: ability soundness changes.

We've talked about being able to do this sort of thing more generally. Like we might compile some pure Unison code to a more efficient builtin, but then the pure Unison source of that definition is still available to look at so people can understand what the code does.

**Comment by ceedubs** [comment]
You've mentioned `bug` and `todo`. Does the scope of this issue also include things like interrupts? Something [that has come up](https://unisonlanguage.slack.com/archives/CLUNF0J5S/p1638803980221100) is a `main` method that should perform some cleanup (connections, file handles, etc) when it is stopped via ctrl-c.

I'll paste some of the Slack discussion in case it goes away.

For example if you run the following program and hit ctrl-c, you'll never see "close".

```
main : '{Exception, IO} ()
main _ =
  open : '{Exception, IO} ()
  open _ = printLine "open"
  close : '{Exception, IO} ()
  close _ = printLine "close"
  go : '{Exception, IO} ()
  go _ =
    io.delay 1000
    !go

  Exception.bracket open close go
```

@dolio has proposed that there could be a builtin `IO.bracket` that would handle async exceptions, but that you'd need a separate `Exception.bracket` or similar to handle ordinary control-flow exceptions.

**Comment by pchiusano** [comment]
Okay we discussed and settled on a new primitive:

```
-- evaluate and catch all exceptions, can be used to implement `bracket` etc
io.tryEval : '{IO} a ->{IO} Either Failure a
```

If it's too hard, do `forkWithErrorHandler` version below.

What to catch?

* Called `bug`
* Called `todo`
* Pattern match failure
* Thread kill?
* Divide by 0
* Everything else just assign to a generic error and dump show into the message of the `Failure`

Other things we considered: 

```
forkWithErrorHandler : '{IO} a -> (Either Failure a ->{IO} ()) ->{IO} ThreadId

bracket : '{IO} r -> (r ->{IO} ()) -> (r ->{IO} a) ->{IO} a
bracket acquire release doTheThing = ... 

-- also considered this, but it's less flexible since the callback is only called
-- forkWithErrorHandler : '{IO} a -> (Failure ->{IO} ()) ->{IO} ThreadId
```

**Comment by ceedubs** [comment]
Fixed by #3365
